### PR TITLE
BaseRNNCell doc changes to make some method docstrings more readable.

### PR DESCRIPTION
**Pull request by rishita**
First pass at some docstring changes in BaseRNNCell. Need to find a good place to hold actual examples. I tried putting examples in the docstrings of methods but without context they are going to be hard to understand.

### CODE REVIEW

**Comment by rishita** [comment]
@szha

File: python/mxnet/rnn/rnn_cell.py [hunk 1]
```diff
@@ -93,11 +93,12 @@ class BaseRNNCell(object):
     Parameters
     ----------
     prefix : str
-        prefix for name of layers
-        (and name of weight if params is None)
+        prefix for names of layers
+        (this prefix is also used for names of weights 
+         if params is None i.e. if params are being created and not reused)
```
Comment (szha) [inline python/mxnet/rnn/rnn_cell.py]: Here's an example
```
`params`
```

**Comment by szha** [comment]
@zackchase

**Comment by piiswrong** [comment]
on examples: we need a page for rnn module. similar to the other parts like module, like docs/rnn.md. Where we describe the basics and show example usage that's too complex for docstring.

**Comment by rishita** [comment]
@piiswrong - Yes, this is also what we decided with @szha. We can add examples on the main page that he is working on currently. 

The docstrings for things like unpack_weights definitely need some context and examples inside would be too long. So I tried to take a midway solution with a small description for now.

**Comment by szha** [comment]
For optional parameters, we should specify the default values

File: python/mxnet/rnn/rnn_cell.py [hunk 2]
```diff
@@ -267,14 +286,14 @@ def unroll(self, length, inputs, begin_state=None, layout='NTC', merge_outputs=N
             If inputs is a list of symbols (usually output of
             previous unroll), they should all have shape
             (batch_size, ...).
-        begin_state : nested list of Symbol
+        begin_state : nested list of Symbol, optional
```
Comment (szha) [inline python/mxnet/rnn/rnn_cell.py]: For optional parameters, we should specify the default values
 Hide all checks

File: python/mxnet/rnn/rnn_cell.py [hunk 3]
```diff
@@ -117,21 +118,29 @@ def reset(self):
...
+        states : list of sym.Variable
+            RNN state from previous step or the output of begin_state().
 
         Returns
         -------
         output : Symbol
-            output symbol
-        states : Symbol
-            state to next step of RNN.
+            output symbol on unrolling this time step.
```
Comment (zackchase) [inline python/mxnet/rnn/rnn_cell.py]: Please captalize "Output". Can you be a little clearer? What does "on unrolling this time step". Do you mean "Symbol corresponding to the output from the RNN on this time step." Might that be better?

File: python/mxnet/rnn/rnn_cell.py [hunk 4]
```diff
@@ -117,21 +118,29 @@ def reset(self):
...
+        states : list of Symbols
+            states symbol is the new state of this RNN after this unrolling.
```
Comment (zackchase) [inline python/mxnet/rnn/rnn_cell.py]: Can truncate to "The new state of this RNN after this  unrolling."

Also can you make "time step" two words?

File: python/mxnet/rnn/rnn_cell.py [hunk 5]
```diff
@@ -117,21 +118,29 @@ def reset(self):
...
+            output symbol on unrolling this time step.
+        states : list of Symbols
+            states symbol is the new state of this RNN after this unrolling. 
+            The type of this is same as the output of begin_state().
+            This can be used as input state to the next timestep 
+            of this RNN.
+
+        See Also
+        --------
+        begin_state: This function can provide the states for the first time step.
```
Comment (zackchase) [inline python/mxnet/rnn/rnn_cell.py]: Do you mean "This function can provide the states for the first time step."

or would it be more accurate and direct to say "This function provides the states for the first time step."?

File: python/mxnet/rnn/rnn_cell.py [hunk 6]
```diff
@@ -117,21 +118,29 @@ def reset(self):
...
+        unroll: This function can be used to unroll the RNN for more than 1 time step.
```
Comment (zackchase) [inline python/mxnet/rnn/rnn_cell.py]: Better to say: 
"unroll: This function unrolls an RNN for a given number of (>=1) time steps." ?

File: python/mxnet/rnn/rnn_cell.py [hunk 7]
```diff
@@ -173,7 +182,7 @@ def begin_state(self, func=symbol.zeros, **kwargs):
         Returns
         -------
         states : nested list of Symbol
-            starting states for first RNN step
+            starting states for the first RNN step
```
Comment (zackchase) [inline python/mxnet/rnn/rnn_cell.py]: Please capitalize "Starting"

File: python/mxnet/rnn/rnn_cell.py [hunk 8]
```diff
@@ -193,19 +202,29 @@ def begin_state(self, func=symbol.zeros, **kwargs):
 
     def unpack_weights(self, args):
         """Unpack fused weight matrices into separate
-        weight matrices
+        weight matrices. 
+
+        For example, say you use a module object mod to run a network that has an lstm cell.
+        In mod.get_params()[0], the lstm parameters are all present as a single big vector.
```
Comment (zackchase) [inline python/mxnet/rnn/rnn_cell.py]: "present as" -> "represented as"

File: python/mxnet/rnn/rnn_cell.py [hunk 9]
```diff
@@ -193,19 +202,29 @@ def begin_state(self, func=symbol.zeros, **kwargs):
 
     def unpack_weights(self, args):
         """Unpack fused weight matrices into separate
-        weight matrices
+        weight matrices. 
+
+        For example, say you use a module object mod to run a network that has an lstm cell.
```
Comment (zackchase) [inline python/mxnet/rnn/rnn_cell.py]: put "mod" in backticks

File: python/mxnet/rnn/rnn_cell.py (hunk 8 above)
Comment (zackchase) [inline python/mxnet/rnn/rnn_cell.py]: Put " mod.get_params()[0]" in double backticks

File: python/mxnet/rnn/rnn_cell.py [hunk 10]
```diff
@@ -193,19 +202,29 @@ def begin_state(self, func=symbol.zeros, **kwargs):
...
+        In mod.get_params()[0], the lstm parameters are all present as a single big vector. 
+        cell.unpack_weights(mod.get_params()[0]) will unpack this vector into a dictionary of
```
Comment (zackchase) [inline python/mxnet/rnn/rnn_cell.py]: Put code in double backticks

File: python/mxnet/rnn/rnn_cell.py [hunk 11]
```diff
@@ -193,19 +202,29 @@ def begin_state(self, func=symbol.zeros, **kwargs):
...
             dictionary containing packed weights.
-            usually from Module.get_output()
+            usually from Module.get_params()[0].
 
         Returns
         -------
         args : dict of str -> NDArray
-            dictionary with weights associated to
-            this cell unpacked.
+            dictionary with unpacked weights associated with
```
Comment (zackchase) [inline python/mxnet/rnn/rnn_cell.py]: capitalize  to "Dictionary"

File: python/mxnet/rnn/rnn_cell.py (hunk 3 above)
Comment (rishita) [inline python/mxnet/rnn/rnn_cell.py]: I think "on this time step" is not explicit enough. I would prefer "Symbol corresponding to the output from the RNN when unrolling for a single time step." (I would expect unroll to be common terminology known to RNN users.)

File: python/mxnet/rnn/rnn_cell.py (hunk 5 above)
Comment (rishita) [inline python/mxnet/rnn/rnn_cell.py]: You can also override the first time step states yourself and not use begin_state(). This is why I used "can". I don't feel too strongly one way or the other on this one though.

**Comment by rishita** [comment]
@zackchase @piiswrong @szha

File: python/mxnet/rnn/rnn_cell.py [hunk 12]
```diff
@@ -284,10 +304,14 @@ def unroll(self, length, inputs, begin_state=None, layout='NTC', merge_outputs=N
 
         Returns
         -------
-        outputs : list of Symbol
-            output symbols.
+        outputs : list of Symbol or Symbol
+            Symbol (if `merge_outputs` is True) or list of Symbols
```
Comment (piiswrong) [inline python/mxnet/rnn/rnn_cell.py]: we should describe the shape of output & state, probably in a separate section like in pytorch
like
(batch_size, length, *) (layout='NTC') or [(batch_size, *)]*length

Comment (rishita) [inline python/mxnet/rnn/rnn_cell.py]: Good idea - Do we want to do this now for this case or go over and fix all of these cases in a separate iteration?

Comment (piiswrong) [inline python/mxnet/rnn/rnn_cell.py]: I'm fine with doing it later.

File: python/mxnet/rnn/rnn_cell.py [hunk 13]
```diff
@@ -117,21 +118,30 @@ def reset(self):
...
 
         Returns
         -------
         output : Symbol
-            output symbol
-        states : Symbol
-            state to next step of RNN.
+            Symbol corresponding to the output from the RNN when unrolling
+            for a single time step.
+        states : nested list of Symbol
```
Comment (szha) [inline python/mxnet/rnn/rnn_cell.py]: should it just be 'list of symbol'?

Comment (rishita) [inline python/mxnet/rnn/rnn_cell.py]: Good point, I just took the type from the begin_state() function and used that everywhere for consistency. But, I think we can make it list of Symbol everywhere including begin_state() itself.

**Pull request merged**